File: cpp/src/leetcode/LC0843GuessTheWord.cpp
#include "leetcode/LC0843GuessTheWord.hpp"

#include <algorithm>
#include <numeric>
#include <ranges>

namespace {
    constexpr uint8_t WORD_LENGTH = 6;
    constexpr uint8_t SCORE_RANGE = WORD_LENGTH + 1;
    constexpr uint8_t MAX_GUESSES = 30;

    std::vector<std::vector<uint8_t>> buildScoreMatrix(const std::vector<std::string> &words);
    size_t findBestCandidate(const std::unordered_set<size_t> &candidates, const std::vector<std::vector<uint8_t>> &scoreMatrix);
    std::unordered_set<size_t> filterCandidates(const std::unordered_set<size_t> &candidates, const std::vector<uint8_t> &guessCandidateScores, int guessScore);
    uint8_t computeMatchCount(const std::string &s1, const std::string &s2);
}

void LC0843GuessTheWord::findSecretWord(const std::vector<std::string>& words, Master& master) {
    const auto scoreMatrix = buildScoreMatrix(words);
    auto candidates = std::views::iota(size_t{0}, words.size()) | std::ranges::to<std::unordered_set>();

    for (uint8_t i = 0; i < MAX_GUESSES; ++i) {
        const size_t candidate = findBestCandidate(candidates, scoreMatrix);

        const int score = master.guess(words[candidate]);
        if (score == WORD_LENGTH) {
            return;
        }

        candidates = filterCandidates(candidates, scoreMatrix[candidate], score);
    }
}

LC0843GuessTheWord::Master::Master(std::string secret, const std::vector<std::string>& words, const uint8_t allowed_guesses) :
        secret(std::move(secret)),
        words(words.begin(), words.end()),
        allowed_guesses(allowed_guesses),
        guessed(false) {}

bool LC0843GuessTheWord::Master::isGuessed() const {
    return guessed;
}

int LC0843GuessTheWord::Master::guess(const std::string& word) {
    if (allowed_guesses == 0) {
        return -1;
    }

    allowed_guesses--;

    if (!words.contains(word)) {
        return -1;
    }

    if (word == secret) {
        guessed = true;
        return WORD_LENGTH;
    }

    return computeMatchCount(secret, word);
}

namespace {
    uint8_t computeMatchCount(const std::string &s1, const std::string &s2) {
        uint8_t matchCount = 0;

        for (size_t i = 0; i < WORD_LENGTH; ++i) {
            if (s1[i] == s2[i])
                matchCount++;
        }

        return matchCount;
    }

    std::vector<std::vector<uint8_t>> buildScoreMatrix(const std::vector<std::string> &words) {
        const size_t N = words.size();

        std::vector scoreMatrix(N, std::vector(N, uint8_t{0}));

        for (size_t i = 0; i < N; ++i) {
            scoreMatrix[i][i] = WORD_LENGTH;

            for (size_t j = i + 1; j < N; ++j) {
                const int matchCount = computeMatchCount(words[i], words[j]);

                scoreMatrix[i][j] = matchCount;
                scoreMatrix[j][i] = matchCount;
            }
        }

        return scoreMatrix;
    }

    std::array<uint8_t, SCORE_RANGE> computePartitionProfile(const std::unordered_set<size_t> &candidates, const std::vector<uint8_t> &scores) {
        std::array<uint8_t, SCORE_RANGE> partitionProfile{};
        for (const auto candidate : candidates) {
            partitionProfile[scores[candidate]]++;
        }
        std::ranges::sort(partitionProfile, std::greater{});
        return partitionProfile;
    }

    bool isBetterPartitionProfile(const std::array<uint8_t, SCORE_RANGE>& currentDistribution, const std::array<uint8_t, SCORE_RANGE>& bestDistribution) {
        for (size_t i = 0; i < SCORE_RANGE; ++i) {
            if (currentDistribution[i] < bestDistribution[i]) {
                return true;
            }
            if (currentDistribution[i] > bestDistribution[i]) {
                return false;
            }
        }

        return false;
    }

    size_t findBestCandidate(const std::unordered_set<size_t> &candidates, const std::vector<std::vector<uint8_t>> &scoreMatrix) {
        if (candidates.size() == 1) {
            return *candidates.begin();
        }
        size_t bestIndex = 0;
        auto bestPartitionProfile = computePartitionProfile(candidates, scoreMatrix[bestIndex]);

        for (size_t i = 1; i < scoreMatrix.size(); ++i) {
            if (const auto partitionProfile = computePartitionProfile(candidates, scoreMatrix[i]);
                isBetterPartitionProfile(partitionProfile, bestPartitionProfile)) {
                bestIndex = i;
                bestPartitionProfile = partitionProfile;
            }
        }

        return bestIndex;
    }

    std::unordered_set<size_t> filterCandidates(const std::unordered_set<size_t> &candidates, const std::vector<uint8_t> &guessCandidateScores, const int guessScore) {
        return candidates |
            std::views::filter([&](const size_t candidate) { return guessCandidateScores[candidate] == guessScore; }) |
            std::ranges::to<std::unordered_set>();
    }
}

File: rust/src/leetcode/lc0843_guess_the_word.rs
use std::collections::HashSet;

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn testcase_1() {
        let secret = "acckzz";
        let words = ["acckzz","ccbazz","eiowzz","abcczz"];
        let allowed_guesses = 10;
        let mut master = Master::new(&words, secret, allowed_guesses);
        find_secret_word(&words, &mut master);
        assert!(master.guessed());
    }

    #[test]
    fn testcase_2() {
        let secret = "hamada";
        let words = ["hamada","khaled"];
        let allowed_guesses = 10;
        let mut master = Master::new(&words, secret, allowed_guesses);
        find_secret_word(&words, &mut master);
        assert!(master.guessed());
    }
}

static WORD_LENGTH: usize = 6;
static SCORE_RANGE: usize = WORD_LENGTH + 1;
static MAX_GUESSES: i8 = 30;

pub fn find_secret_word(words: &[&str], master: &mut Master) {

    let score_matrix = build_score_matrix(words);
    let mut candidates: HashSet<usize> = (0..words.len()).collect();

    for _ in 0..MAX_GUESSES {
        let guess_index = find_best_candidate(&candidates, &score_matrix);

        let score = master.guess(words[guess_index]);
        if score == WORD_LENGTH as i8 {
            return;
        }

        candidates = filter_candidates(candidates, &score_matrix[guess_index], score);
    }
}

fn filter_candidates(candidates: HashSet<usize>, guess_candidate_scores: &Vec<i8>, guess_score: i8) -> HashSet<usize> {
    candidates.into_iter()
        .filter(|&candidate| { guess_candidate_scores[candidate] == guess_score })
        .collect()
}

fn find_best_candidate(candidates: &HashSet<usize>, score_matrix: &Vec<Vec<i8>>) -> usize {
    if candidates.len() == 1 {
        return candidates.iter().next().unwrap().clone();
    }

    let mut best_index = 0;
    let mut best_score_frequency = compute_score_frequency(candidates, &score_matrix[best_index]);

    for i in 1..score_matrix.len() {
        let score_frequency = compute_score_frequency(candidates, &score_matrix[i]);
        if is_better_score_frequency(score_frequency, best_score_frequency) {
            best_index = i;
            best_score_frequency = score_frequency;
        }
    }

    best_index
}

fn compute_score_frequency(candidates: &HashSet<usize>, score_matrix: &Vec<i8>) -> [usize; SCORE_RANGE] {
    let mut score_frequency = [0; SCORE_RANGE];

    for &candidate in candidates {
        score_frequency[score_matrix[candidate] as usize] += 1;
    }

    score_frequency.sort_by(|a, b| b.cmp(a));

    score_frequency
}

fn is_better_score_frequency(score_frequency1: [usize; SCORE_RANGE], score_frequency2: [usize; SCORE_RANGE]) -> bool {
    for i in 0..SCORE_RANGE {
        if score_frequency1[i] < score_frequency2[i] {
            return true;
        }
        if score_frequency1[i] > score_frequency2[i] {
            return false;
        }
    }

    false
}

fn build_score_matrix(words: &[&str]) -> Vec<Vec<i8>> {
    let mut score_matrix = vec![vec![0i8; words.len()]; words.len()];
    for i in 0..words.len() {
        score_matrix[i][i] = WORD_LENGTH as i8;
        for j in i + 1..words.len() {
            let match_score = compute_match_count(words[i], words[j]);
            score_matrix[i][j] = match_score;
            score_matrix[j][i] = match_score;
        }
    }
    score_matrix
}

fn compute_match_count(word1: &str, word2: &str) -> i8 {
    word1.chars().zip(word2.chars())
    .filter(|&(c1, c2)| c1 == c2)
    .count() as i8
}

pub struct Master {
    words: HashSet<String>,
    secret: String,
    allowed_guesses: u8,
    guessed: bool
}

impl Master {
    pub fn guess(&mut self, word: &str) -> i8 {
        if self.allowed_guesses == 0 {
            return -1;
        }

        self.allowed_guesses -= 1;

        if !self.words.contains(word) {
            return -1;
        }

        if word == self.secret {
            self.guessed = true;
            return WORD_LENGTH as i8;
        }

        compute_match_count(word, self.secret.as_str())
    }
}

#[cfg(test)]
impl Master {
    fn new(words: &[&str], secret: &str, allowed_guesses: u8) -> Self {
        Self {
            words: words.iter().map(|s| s.to_string()).collect(),
            secret: secret.to_string(),
            allowed_guesses,
            guessed: false,
        }
    }

    fn guessed(&self) -> bool {
        self.guessed
    }
}
